// util-format/lib/index.js
var moment = require('moment');

function date(_date, format, new_format) { //日期格式化
	format = format || 'YYYY-MM-DD'; //解决没传的兼容性
	if ((typeof _date) == 'number') {
		return new moment().add(_date, 'days').format(format);
	} else {
		var _new = new moment(_date, format);
		return _new.format(new_format || 'MM-DD HH:mm');
	}
}

function currency(num, dot) { //格式化货币 000,000,000方式显示
	if (dot === 0) {
		num = Number(num).toFixed(0);
	} else {
		dot = dot || 2;
		num = Number(num).toFixed(dot);
	}
	var n = '' + num;
	var r = n.replace(/(^|-|\s)\d+(?=\.?\d*($|\s))/g, function(m) {
		return m.replace(/(?=(?!\b)(\d{3})+$)/g, ',');
	});
	if (dot === 0) {
		return r;
	} else {
		var i = r.indexOf('.');
		if (i >= 0) return r.substring(0, i + 1 + dot);
		else {
			r += '.';
			for (var i = dot; i > 0; dot--) {
				r += '0';
			};
			return r;
		}
	}
}

function option(str, options) {
	var start = options.indexOf(str + '-');
	if (start < 0)
		return '';
	var end = options.indexOf(',', start);
	if (end < 0) {
		end = options.length;
	}
	return options.substring((start + 1 + ('' + str).length), end);
}

function string_max(str, length) {
	return str.length > length ? (str.substring(0, length) + '...') : str;
}

function mask(text, start, length) {
	if (text) {
		var m = '';
		for (var i = 0; i < length; i++) {
			m += '*';
		}
		return text.substr(0, start) + m + text.substr(start + length);
	}
	return '';
}

function mask_mobile(text) {
	return mask(text, 2, 5);
}

function mask_name(text) {
	return mask(text, 1, text.length - 1);
}

function mask_bankcard(text) {
	return mask(text, 4, text.length - 8);
}

module.exports = {
	"date": date,
	"currency": currency,
	"option": option,
	"string_max": string_max,
	"mask": mask,
	"mask_mobile": mask_mobile,
	"mask_name": mask_name,
	"mask_bankcard": mask_bankcard
};
